### PR TITLE
feat(std): Add `std/url` module.

### LINKED ISSUE #3515: Add std/url with URL utilities
Should have similar functionality to path: parent, basename, ext

but operate on JS native URL objects

### PR DESCRIPTION
**Pull request by tr1ckydev**
Utilities for working with URL paths (#3515).

closes #3515

### CODE REVIEW

File: url/dirname_test.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  ],
+  [
+    new URL("https://deno.land/std/assert/mod.ts"),
+    new URL("https://deno.land/std/assert"),
+  ],
+  [
+    new URL("https://deno.land/std/assert/mod.ts?foo=bar"),
+    new URL("https://deno.land/std/assert"),
+  ],
+  [new URL("https://deno.land///"), new URL("https://deno.land")],
```
Comment (lino-levan) [inline url/dirname_test.ts:19]: You should probably include a test with a hash in it

File: url/extname_test.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+import { assertEquals } from "../assert/mod.ts";
+import * as url from "./mod.ts";
+
+const TESTSUITE = [
+  ["https://deno.land/std/assert/mod.ts", ".ts"],
+  [new URL("https://deno.land/std/assert/mod.ts"), ".ts"],
+  [new URL("https://deno.land/std/assert/mod.ts?foo=bar"), ".ts"],
+  [new URL("https://deno.land/std/assert/mod."), "."],
+  [new URL("https://deno.land/std/assert/mod"), ""],
```
Comment (lino-levan) [inline url/extname_test.ts:11]: Again, hash test would be nice.

File: url/join_test.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    new URL("https://deno.land/std/assert/mod.ts"),
+  ],
+  [["https://deno.land/"], new URL("https://deno.land/")],
+];
+
+Deno.test("join", function () {
+  for (const [[test_url, ...paths], expected] of TESTSUITE) {
+    assertEquals(url.join(test_url, ...paths), expected);
+  }
+});
```
Comment (lino-levan) [inline url/join_test.ts:26]: Now that I'm looking at the tests, I'm not sure if join even makes sense.

File: url/normalize.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+import { posixNormalize } from "../path/_normalize.ts";
+
+/**
+ * Normalize the `url`, resolving mutiple `'/'`s into `'//'` after protocol and remaining into `'/'`.
```
Comment (lino-levan) [inline url/normalize.ts:7]: I'd expect normalize to normalize away `..` -- Is this the case?

File: url/dirname_test.ts (hunk 1 above)
Comment (tr1ckydev) [reply url/dirname_test.ts:23]: Added now.

File: url/extname_test.ts (hunk 2 above)
Comment (tr1ckydev) [reply url/extname_test.ts:12]: Added.

File: url/join_test.ts (hunk 3 above)
Comment (tr1ckydev) [reply url/join_test.ts:26]: Why is it so?

Comment (lino-levan) [reply url/join_test.ts:26]: I guess it handles edgecases around double `/` so maybe it's still useful. Not entirely sure what the utility is. Thoughts @marvinhagemeister?

File: url/normalize.ts (hunk 4 above)
Comment (tr1ckydev) [reply url/normalize.ts:7]: Can you provide me with a test case?

Comment (lino-levan) [reply url/normalize.ts:7]: Something like
```ts
  [
    "https://deno.land/std/assert/../async/retry.ts",
    new URL("https://deno.land/std/async/retry.ts"),
  ],
  ```
  I'd also probably test that `https://deno.land/..` normalizes to `https://deno.land/`

Comment (tr1ckydev) [reply url/normalize.ts:7]: Both the cases pass the test.

Comment (tr1ckydev) [reply url/normalize.ts:7]: Added the cases.

Comment (lino-levan) [reply url/normalize.ts:7]: Could you add that it handles this to the JSDoc?

Comment (tr1ckydev) [reply url/normalize.ts]: Added.

**Comment by tr1ckydev** [comment]
Should `common` from std/path be implemented for url? Need suggestion.

**Comment by lino-levan** [comment]
I'm trying to think of a usecase and can't come up with one off of the top of my head. I think a lot of these things can be done as a follow-up PR. So no pressure to get everything in right now!

**Comment by tr1ckydev** [comment]
I have fixed couple of bugs and added new edge cases tests. Will be adding them later by today.

**Comment by timreichen** [comment]
What would you think of adding a `isUrl(input: string)` function, either in this PR or the future?
Seems like both [isurl](https://www.npmjs.com/package/isurl) and [is-url](https://www.npmjs.com/package/is-url) have millions of weekly downloads.

**Comment by 0f-0b** [comment]
> What would you think of adding a `isUrl(input: string)` function, either in this PR or the future?

There is [`URL.canParse`](https://developer.mozilla.org/docs/Web/API/URL/canParse_static), although browser support is currently poor.

**Comment by tr1ckydev** [comment]
https://github.com/tr1ckydev/pathurl
This is my package with all the proposed path functions for URLs along with various tests. If it's acceptable then I will port all the functions from here into `std/url`.

File: url/normalize.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+// This module is browser compatible.
+
+import { posixNormalize } from "../path/_normalize.ts";
+
+/**
+ * Normalize the `url`, resolving `'..'` and `'.'` segments and mutiple `'/'`s into `'//'` after protocol and remaining into `'/'`.
```
Comment (kt3k) [inline url/normalize.ts:7]:
```suggestion
 * Normalize the `url`, resolving `'..'` and `'.'` segments and multiple `'/'`s into `'//'` after protocol and remaining into `'/'`.
```

**Comment by kt3k** [comment]
`dirname` might sound strange because it returns a URL instead of a string (it doesn't feel like 'name' to me), but this looks good enough for first pass.

**Review by kt3k — APPROVED** [review]
@tr1ckydev LGTM. Nice work!

**Comment by tr1ckydev** [comment]
> `dirname` might sound strange because it returns a URL instead of a string (it doesn't feel like 'name' to me), but this looks good enough for first pass.

Agree, but as it's from url module so it can be expected. Even though dirname from path module returns parent path so works in case of url too imo.
